src/librustc/middle/ty/mod.rs

```diff
@@ -2396,6 +2396,45 @@ impl<'tcx> ctxt<'tcx> {
             || self.sess.cstore.item_super_predicates(self, did))
     }
 
+    /// If `type_needs_drop` returns true, then `ty` is definitely
+    /// non-copy and *might* have a destructor attached; if it returns
+    /// false, then `ty` definitely has no destructor (i.e. no drop glue).
+    ///
+    /// (Note that this implies that if `ty` has a destructor attached,
+    /// then `type_needs_drop` will definitely return `true` for `ty`.)
+    pub fn type_needs_drop(&self, ty: Ty<'tcx>) -> bool {
+        self.type_needs_drop_given_env(ty, &self.empty_parameter_environment())
```

Comment:
Is it really fine to use empty param environment here? I assume this is supposed to hold some sort of substs for generics. It was fine to do pass empty environment in trans, because all types are monomorphized there, but in `middle::ty` or `mir::hair::cx` you cannot assume that types will be monomorphized (they usually won’t be).